### PR TITLE
Add `Set<int>` and related operations as index sort

### PR DESCRIPTION
**Pull request by ranjitjhala**
See [list01.rs](flux-tests/tests/pos/enums/list01.rs)

### CODE REVIEW

File: flux-middle/src/fhir.rs [hunk 1]
```diff
@@ -407,6 +407,18 @@ newtype_index! {
...
+pub enum SortCtor {
+    Set,
+    User { name: Symbol, arity: usize },
+}
+
+impl SortCtor {
+    pub fn set() -> Self {
+        SortCtor::Set
+    }
+}
```
Comment (nilehmann) [inline flux-middle/src/fhir.rs:420]: I'd remove this, writing `SortCtor::Set` sounds better than `SortCtor::set()`

```suggestion
```

File: flux-desugar/src/desugar.rs [hunk 2]
```diff
@@ -865,27 +865,65 @@ fn resolve_func_sort(
...
+        Err(sess.emit_err(errors::UnresolvedSort::new(ident)))
+    }
+}
+
+fn ctor_arity(ctor: &fhir::SortCtor) -> usize {
+    match ctor {
+        fhir::SortCtor::Set => 1,
+        fhir::SortCtor::User { arity, .. } => *arity,
+    }
+}
```
Comment (nilehmann) [inline flux-desugar/src/desugar.rs:902]: I'd make this as a method of `SortCtor`

```rust
impl SortCtr {
    pub fn arity(&self) -> usize {
        match self {
            SortCtr::Set => 1,
            SortCtr::User { arity, ..} => arity,
        }
    }
}

File: flux-middle/src/fhir.rs [hunk 3]
```diff
@@ -605,7 +619,7 @@ pub struct FuncDecl {
 #[derive(Debug, Clone, Copy, TyEncodable, TyDecodable, PartialEq, Eq, Hash)]
 pub enum FuncKind {
     /// Theory symbols "interpreted" by the SMT solver
-    Thy,
+    Thy(Symbol),
```
Comment (nilehmann) [inline flux-middle/src/fhir.rs:622]: Can you add a comment saying this is the name in fixpoint?

File: flux-middle/src/fhir.rs [hunk 4]
```diff
@@ -892,25 +910,76 @@ impl Map {
...
+            Symbol::intern("Set_cup"),
+            vec![Sort::set(Sort::Int), Sort::set(Sort::Int)],
+            Sort::set(Sort::Int),
+        );
+        // self.insert_theory_func(
+        //     Symbol::intern("set_is_empty"),
+        //     Symbol::intern("Set_is_empty")
+        //     vec![Sort::set(Sort::Int)],
+        //     Sort::Bool,
+        // );
```
Comment (nilehmann) [inline flux-middle/src/fhir.rs:976]:
```suggestion
```

File: flux-refineck/src/lib.rs [hunk 5]
```diff
@@ -76,6 +76,7 @@ pub fn check_fn(
 
         // PHASE 3: invoke fixpoint on the constraint
         refine_tree.simplify();
+
```
Comment (nilehmann) [inline flux-refineck/src/lib.rs:79]:
```suggestion
```

File: flux-tests/tests/pos/enums/list01.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+#[flux::sig(fn(&List[@xs]) -> bool[xs == set_emp()])]
+pub fn is_empty(l: &List) -> bool {
+    match l {
+        List::Nil => true,
+        List::Cons(_, _) => false,
+    }
+}
+
+// WHY DOES THIS FAIL?
```
Comment (nilehmann) [inline flux-tests/tests/pos/enums/list01.rs:28]: I think this is not failing anymore, no?

File: flux-tests/tests/pos/enums/list01.rs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+// #[flux::sig(fn () -> List[set_emp()])]
+pub fn null() -> List {
+    List::Nil
+}
+
+#[flux::sig(fn(i32{v: false}) -> T)]
+pub fn never<T>(_: i32) -> T {
+    loop {}
+}
```
Comment (nilehmann) [inline flux-tests/tests/pos/enums/list01.rs:39]: I typically call this `unreachable` like the rust macro `unreachable!` and define it like so

```suggestion
#[flux::sig(fn() -> T requires false)]
pub fn unreachable<T>() -> T {
    unreachable!()
}
```

File: flux-tests/tests/pos/enums/list00.rs [hunk 8]
```diff
@@ -15,7 +19,7 @@ pub enum List {
     Cons(i32, Box<List>),
 }
 
-#[flux::sig(fn(&List[@n]) -> bool[n == 0])]
+#[flux::sig(fn(&List[@n]) -> bool[is_zero(n)])]
```
Comment (nilehmann) [inline flux-tests/tests/pos/enums/list00.rs:22]: Is there any reason we added this?

**Review by nilehmann — APPROVED** [review]
LGTM. A couple of minor comments


Also, I just remember that we already have user-defined "opaque" sorts. We should probably unify those with sort constructors. That can come in another PR.

File: flux-middle/src/fhir.rs (hunk 1 above)
Comment (ranjitjhala) [reply flux-middle/src/fhir.rs:420]: done!

File: flux-tests/tests/pos/enums/list00.rs (hunk 8 above)
Comment (ranjitjhala) [reply flux-tests/tests/pos/enums/list00.rs:22]: oops debugging detritus, let me undo.

**Pull request merged**